Specify null safety subtyping

eernstg opened the pull request:
This PR contains the null-safety updates from #2605 (the overall null-safety PR) that are concerned with subtyping. The intention is that we process and land this PR, and #2605 will then be made smaller when it is rebased over this one, thus allowing us to deal with the null-safety update in smaller portions.

eernstg committed:
Integrate the null-safety subtyping rules into dartLangSpec.tex

eernstg committed:
Integrated null-safety into the appendix about subtyping

eernstg committed:
Fix a couple of typos

lrhn commented on specification/dartLangSpec.tex:
```diff
@@ -23130,156 +23146,238 @@ \subsection{Operator Precedence}
-which means that this seemingly expensive step can be confined to some extent.%
-}
+  \commentary{%
+    Note that this rule is admissible, and can be safely elided if desired.%
```
Define "admissible". It's a term of art not used elsewhere in the specification, so we cannot assume the reader to be familiar with it.

lrhn reviewed (commented):
Can't really review this as source.
Do we still have a way to generate a rendering showing the changes?

eernstg committed:
Adjust dart.sty to change like specify_null_safety_sep21

eernstg committed:
WIP

eernstg committed:
Align this PR with specify_null_safety_sep21

eernstg committed:
WIP

eernstg commented:
[dartLangSpec-3515.pdf](https://github.com/user-attachments/files/17414000/dartLangSpec-3515.pdf)

eernstg commented:
Based on the output from `git diff main...`, here is an overview of the changes:

### Changed locations

#### Front matter

Only comments.

#### @@ -21633,128 +21631,188 @@

Section 'Subtypes': Now mentions intersection types.

Figure 'Subtype rules': Changed a lot.

Subsection 'Meta-Variables': Slightly changes presentation of
metavariables.

Subsection 'Subtype Rules': Introduces the 'canonical syntax' of a
type. Controversial! An alternative could be to say that subtype rules
are _not_ applicable to syntax, they are applicable to semantic
entities that are normalized in the same way as canonical syntax does
it. Extensively rewritten.

Subsection 'Being a Subtype': Extensively rewritten.

Subsubsection 'Informal subtype rule descriptions': Rewritten extensively.

Subsection 'Additional Subtyping Concepts': Extensively rewritten.

Section 'Function Types': Last paragraph modified.

Section 'Type Function': Rewritten.

#### @@ -22482,7 +22498,7 @@

Just the title of section 'Type dynamic' was adjusted, for consistency
(that is: titles do not use non-standard fonts).

Section 'Type FutureOr': Just adding missing `\Error{}` on static
errors, and updating `Object` to `Object?` in an example.

Section 'Type void': Removed commentary that reiterated an obsolete
definition of `Type` equality. Removed the item that enabled `await
expressionOfTypeVoid`, because that rule that enabled this usage has
been removed from the language. Example at the end is updated.

Subsection 'Void Soundness': Example updated, plus body text about the
example.

#### @@ -23324,156 +23324,238 @@

Section 'Appendix: Algorithmic Subtyping': Updated to use the
NNBD-updated presentation of the subtype rules.

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg commented:
I tried to use `latexdiff` to generate a PDF that shows the differences, but it doesn't handle "non-standard" LaTeX (which seems to be difficult to avoid). I can generate a PDF that shows some differences and stops at page 232, but it would take some more experimentation to produce something that shows the differences in a way that can be relied upon.

eernstg committed:
Added section about explicitly resolved (fka canonical) syntax

eernstg commented:
I added a section about 'being the same type', which means that the thing previously known as 'canonical syntax' (now renamed to 'explicitly resolved syntax') is encapsulated in this section.

eernstg committed:
Clean up whitespace

eernstg committed:
Whitespace

eernstg committed:
Whitespace

eernstg committed:
WIP

eernstg committed:
Rename type variables

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg reviewed (commented):
Added a couple of comments about section references which will be resolved in a separate PR where those sections are added.

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg committed:
WIP

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -21839,128 +21839,188 @@ \subsection{Subtypes}
+    % TODO(eernst): Comment out this rule as redundant and
+    % renumber, it is an instance of \SrnRightFutureOrB.
+    \RuleRaw{\SrnRightObjectFour}{%
+      \mbox{$S$ is an interface type or \FUNCTION}}{S}{\code{Object}}
```
Yes. `Record` is not in the language at 2.13, but I added a TODO to handle this. I also removed the TODO which was there previously because it applies to `\SrnNullTwo` rather than `\SrnRightObjectFour`.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -21839,128 +21839,188 @@ \subsection{Subtypes}
+      \mbox{$S$ is an interface type or \FUNCTION}}{S}{\code{Object}}
+    % ------------------------------------------------ Left Null 2
+    \Rule{\SrnNullTwo}{\code{Null}}{T}{%
+      \code{Null}}{\code{FutureOr<$T$>}}
```
Exactly, done.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -22025,41 +22087,29 @@ \subsubsection{Subtype Rules}
+\LMHash{}%
+Every \synt{typeName} used in a type mentioned
+in section~\ref{subtypes} and its subsections
+is assumed to have no compile-time error,
```
Yes, done.

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -22489,33 +22489,28 @@ \subsection{Function Types}
 }
 
 \LMHash{}%
 A function object is always an instance of some class $C$ that implements
```
Done! No need to have weasel talk that doesn't even match the reality.

eernstg committed:
WIP

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -21839,128 +21839,188 @@ \subsection{Subtypes}
+    \RuleTwo{\SrnLeftNullable}{S}{T}{\code{Null}}{T}{%
+      \code{$S$?}}{T}
+    % ------------------------------------------------ Left Promoted Variable A
+    \Axiom{\SrnLeftPromotedVariableOne}{X \& S}{X}
```
Agreed, we need the more general form because we don't have a general transitivity rule. Done.

eernstg committed:
Adjust SrnLeftPromotedVariableOne to avoid implied transitivity

eernstg committed:
Update textual presentation of SrnLeftPromotedVariableOne

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -23529,157 +23668,266 @@ \subsection{Operator Precedence}
+  if $S$ is \code{Null} then:
+  \begin{itemize}
+  \item
+    if $T$ is a type variable (promoted or not) the query is false.
```
Optimization is definitely one of the main reasons for having this ruleset &mdash; it is intended to show how Dart subtyping can be decided with good performance. The inference rules in figure 4 are intended to specify the intended relation with no concerns about decision speed. I think that's OK.

eernstg committed:
WIP

eernstg commented on specification/dartLangSpec.tex:
```diff
@@ -23529,157 +23668,266 @@ \subsection{Operator Precedence}
+\item
+  If $S$ is \code{($V_0$, \ldots, $V_n$, \{$V_{n+1} d_{n+1}$, \ldots, $V_m d_m$\})}
+  and $T$ is \code{($S_0$, \ldots, $S_n$, \{$S_{n+1} d_{n+1}$, \ldots, $S_m d_m$\})}
+  then \SubtypeNE{S}{T} if{}f \SubtypeNE{V_i}{S_i} for each $i \in 0 .. m$.
```
Done.

chloestefantsova reviewed (approved):
lgtm :)
